snow/consensus/snowman/poll/early_term_no_traversal.go

```diff
@@ -192,15 +197,63 @@ func (p *earlyTermNoTraversalPoll) Finished() bool {
 		return true // Case 2
 	}
 
-	_, freq := p.votes.Mode()
-	if freq >= p.alphaPreference && maxPossibleVotes < p.alphaConfidence {
+	//    v
+	//   /
+	//  u
+	// We build a vote graph where each vertex represents a block ID.
+	// A vertex 'v' is a parent of vertex 'u' if the ID of 'u' corresponds
+	// to a block that is the successive block of the corresponding block for 'v'.
+	votesGraph := buildVoteGraph(p.bt.GetParent, p.votes)
+
+	// If vertex 'v' is a parent of vertex 'u', then a vote for the ID of vertex 'u'
+	// should also be considered as a vote for the ID of the vertex 'v'.
+	transitiveVotes := computeTransitiveVoteCountGraph(&votesGraph, p.votes)
+
+	//     v
+	//   /   \
+	//  u     w
+	// If two competing blocks 'u', 'w' are potential successors to a block 'v',
+	// snowman would instantiate a unary snowflake instance on the prefix of 'u' and 'w'.
+	// The prefix inherits the votes for the IDs of 'u' and 'w'.
+	// We therefore compute the transitive votes for all prefixes of IDs
+	// for each bifurcation in the transitive vote graph.
+	transitiveVotesForPrefixes := computeTransitiveVotesForPrefixes(&votesGraph, transitiveVotes)
+
+	// We wish to compute the votes for snowflake instances, no matter if they correspond to an actual block ID,
+	// or a unary snowflake instance for a shared prefix between a bifurcation of two competing blocks.
+	// For that, only the number of votes and existence of such snowflake instances matters.
+	voteCountsForIDsOrPrefixes := aggregateVotesFromPrefixesAndIDs(transitiveVotesForPrefixes, transitiveVotes)
+
+	// Given the aforementioned votes, we wish to see whether there exists a snowflake instance
+	// that can benefit from waiting for more invocations of Vote().
+	// We therefore check each amount of votes separately and see if voting for that snowflake instance
+	// should terminate, as it cannot be improved by further voting.
+	weCantImproveVoteForSomeIDOrPrefix := make(booleans, len(voteCountsForIDsOrPrefixes))
+	for i, completedVotes := range voteCountsForIDsOrPrefixes {
+		shouldTerminate := p.shouldTerminateDueToConfidence(completedVotes, remaining)
+		weCantImproveVoteForSomeIDOrPrefix[i] = shouldTerminate
+	}
+
+	// We should terminate the poll only when voting for all snowflake instances should terminate.
+	if weCantImproveVoteForSomeIDOrPrefix.allTrue() && len(weCantImproveVoteForSomeIDOrPrefix) > 0 {
 		p.finished = true
+	}
```

Comment:
Why do we add this `booleans` type? It doesn't seem to clarify the code imo
```suggestion
	shouldTerminate := len(voteCountsForIDsOrPrefixes) > 0
	for _, completedVotes := range voteCountsForIDsOrPrefixes {
		shouldTerminate = shouldTerminate && p.shouldTerminateDueToConfidence(completedVotes, remaining)
	}

	// We can terminate the poll if all snowflake instances should terminate.
	if shouldTerminate {
		p.finished = true
	}
```

Reply:
I thought it gives clarity.

Addressed [here](https://github.com/ava-labs/avalanchego/pull/3337/commits/862d10c6c342a1dfb92369b55e983399a0ec943f).